pkg/controller/controller.go

```diff
@@ -997,6 +1070,150 @@ func (p *csiProvisioner) Delete(volume *v1.PersistentVolume) error {
 	return err
 }
 
+func (p *csiClaimController) runClaimWorker() {
+	for p.processNextClaimWorkItem() {
+	}
+}
+
+// processNextClaimWorkItem processes items from claimQueue
+func (p *csiClaimController) processNextClaimWorkItem() bool {
+	obj, shutdown := p.claimQueue.Get()
+	if shutdown {
+		return false
+	}
+
+	err := func(obj interface{}) error {
+		defer p.claimQueue.Done(obj)
+		var key string
+		var ok bool
+		if key, ok = obj.(string); !ok {
+			p.claimQueue.Forget(obj)
+			return fmt.Errorf("expected string in workqueue but got %#v", obj)
+		}
+
+		if err := p.syncClaimHandler(key); err != nil {
+			klog.Warningf("Retrying syncing claim %q after %v failures", key, p.claimQueue.NumRequeues(obj))
+			p.claimQueue.AddRateLimited(obj)
+		}
+
+		p.claimQueue.Forget(obj)
+		return nil
+	}(obj)
+
+	if err != nil {
+		utilruntime.HandleError(err)
+		return true
+	}
+
+	return true
+}
+
+// enqueueClaimUpadate takes an obj and converts it into UID that is then put onto claim work queue.
+func (p *csiClaimController) enqueueClaimUpadate(obj interface{}) {
+	new, ok := obj.(*v1.PersistentVolumeClaim)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("expected claim but got %+v", new))
+		return
+	}
+
+	// Timestamp didn't appear
+	if new.DeletionTimestamp == nil {
+		return
+	}
+
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(err)
+		return
+	}
+
+	p.claimQueue.Add(key)
+}
+
+// syncClaimHandler gets the claim from informer's cache then calls syncClaim
+func (p *csiClaimController) syncClaimHandler(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	claimObj, err := p.claimLister.PersistentVolumeClaims(namespace).Get(name)
+	if err != nil {
+		if apierrs.IsNotFound(err) {
+			utilruntime.HandleError(fmt.Errorf("Item '%s' in work queue no longer exists", key))
+			return nil
+		}
+
+		return err
+	}
+
+	return p.syncClaim(claimObj)
+}
+
+func (p *csiClaimController) syncClaim(obj interface{}) error {
+	claim, ok := obj.(*v1.PersistentVolumeClaim)
+	if !ok {
+		return fmt.Errorf("expected claim but got %+v", obj)
+	}
+
+	if !checkFinalizer(claim, pvcCloneFinalizer) {
+		return nil
+	}
+
+	pvcList, err := p.client.CoreV1().PersistentVolumeClaims("").List(metav1.ListOptions{})
```

Comment:
IMo you should use informer here, we don't want to load the API server with every PVC being cloned.

Reply:
Fixed.